applications/PfemFluidDynamicsApplication/custom_processes/update_conditions_on_free_surface_process.hpp

```diff
@@ -203,37 +216,56 @@ class UpdateConditionsOnFreeSurfaceProcess : public Process {
 								}
 							}
 
-							if (number_of_rigid_nodes != number_of_nodes_in_face) {
-
-								Condition::Pointer p_condition = r_reference_condition.Create(++condition_id, face_nodes, p_property);
-
+							// Check if it is a free surface
+							if (number_of_rigid_nodes != number_of_nodes_in_face)
+							{
+								// Create new condition on free surface
+								Condition::Pointer p_condition = r_reference_condition.Create(condition_id++, face_nodes, p_property);
 								mrModelPart.Conditions().push_back(p_condition);
 								r_sub_model_part.Conditions().push_back(p_condition);
+
+								// Add nodes to free surface
+								for (unsigned int j = 1; j <= number_of_nodes_in_face; ++j) {
+									bool add = true;
+									for (auto i_node(r_sub_model_part.NodesBegin()); i_node != r_sub_model_part.NodesEnd(); ++i_node)
+										if (i_node->Id() == r_geometry(lpofa(j, iface))->Id())
+											add = false;
+									if (add)
+										r_sub_model_part.Nodes().push_back(r_geometry(lpofa(j, iface)));
+								}
 							}
 						}
 						iface += 1;
 					}
 				}
 				KRATOS_INFO_IF("UpdateConditionsAfterRemeshProcess ", mEchoLevel > 0)
 				    << " SubModelPart: " << mListOfSubModelParts[i] << " Rebuild time: " << time_elapsed.ElapsedSeconds() << std::endl;
+			}
 
-			} else {
+			// For the other sub model parts we clone the nodes/conditions from the first sub model part
+			else
+			{
 				BuiltinTimer time_elapsed;
-				// For the other sub model parts we clone the conditions from the first sub model part
+
+				// Get first free surface sub model part
 				ModelPart& r_origin_part = mrModelPart.GetSubModelPart(mListOfSubModelParts[0]);
 
+				// Copy nodes from first free surface sub model part
+				r_sub_model_part.AddNodes(r_origin_part.NodesBegin(), r_origin_part.NodesEnd());
+
+				// Initialize vector of new conditions
 				ModelPart::ConditionsContainerType temp_conditions;
 				temp_conditions.reserve(r_origin_part.NumberOfConditions());
 
-				for (auto i_cond = r_origin_part.ConditionsBegin(); i_cond != r_origin_part.ConditionsEnd(); ++i_cond) {
-					Properties::Pointer properties = i_cond->pGetProperties();
-
-					// Reuse the geometry of the old element (to save memory)
-					Condition::Pointer p_condition = r_reference_condition.Create(++condition_id, i_cond->pGetGeometry(), properties);
-
+				// Loop over conditions created for first free surface sub model part
+				for (auto i_cond = r_origin_part.ConditionsBegin(); i_cond != r_origin_part.ConditionsEnd(); ++i_cond)
+				{
```

Comment:
be careful with the "style" guide in c++, I'll tell you a few tips for the future